lib/ansible/modules/cloud/vmware/vmware_export_ovf.py

```diff
@@ -0,0 +1,332 @@
+                              total_bytes_to_write):
+        url_hostname = self.params['hostname']
+        mf_content = 'SHA256(' + os.path.basename(temp_target_disk) + ')= '
+        sha256_hash = hashlib.sha256()
+        if '*' in device_url:
+            device_url = device_url.replace('*', url_hostname)
+        with open(self.mf_file, 'a') as mf_handle:
+            with open(temp_target_disk, 'wb') as handle:
+                try:
+                    response = requests.get(device_url, stream=True, headers=headers, cookies=cookies, verify=False)
+                except Exception as err:
+                    lease_updater.httpNfcLease.HttpNfcLeaseAbort()
+                    lease_updater.stop()
+                    self.module.fail_json(msg='Exception caught when getting %s, %s' % (device_url, to_text(err)))
+                if not response.ok:
+                    lease_updater.httpNfcLease.HttpNfcLeaseAbort()
+                    lease_updater.stop()
+                    self.module.fail_json(msg='Getting %s return code %d' % (device_url, response.status_code))
+                current_bytes_written = 0
+                for block in response.iter_content(chunk_size=self.chunk_size):
+                    if block:
+                        handle.write(block)
+                        sha256_hash.update(block)
+                        handle.flush()
+                        os.fsync(handle.fileno())
+                    current_bytes_written += len(block)
+                    written_percent = ((current_bytes_written + total_bytes_written) * 100) / total_bytes_to_write
+                    lease_updater.progressPercent = int(written_percent)
+            mf_handle.write(mf_content + sha256_hash.hexdigest() + '\n')
+        self.facts['device_files'].append(temp_target_disk)
+        return current_bytes_written
+
+    def export_to_ovf_files(self):
+        export_with_iso = False
+        if 'export_with_images' in self.params and self.params['export_with_images']:
+            export_with_iso = True
+        nvram_file_name = ''
+        ovf_files = []
+        # get http nfc lease firstly
+        http_nfc_lease = self.vm.ExportVm()
+        # create a thread to track file download progress
+        lease_updater = LeaseProgressUpdater(http_nfc_lease, self.lease_interval)
+        total_bytes_written = 0
+        # total storage space occupied by the virtual machine across all datastores
+        total_bytes_to_write = self.vm.summary.storage.unshared
+        # new deployed VM with no OS installed
+        if total_bytes_to_write == 0:
+            total_bytes_to_write = self.vm.summary.storage.committed
+            if total_bytes_to_write == 0:
+                http_nfc_lease.HttpNfcLeaseAbort()
+                self.module.fail_json(msg='Total storage space occupied by the VM is 0.')
+        headers = {'Accept': 'application/x-vnd.vmware-streamVmdk'}
+        cookies = self.get_cookies()
+        lease_updater.start()
```

Comment:
It looks like there's enough to the lease, deviceUrl, and downloading file(s) that I would recomment a module_utils/vmware_lease.py or module_utils/vmware_download.py be created where the download and lease logic are make available for other modules to use.  For example these leases are used in vmware_deploy_ovf too.

Although the *Rule of Three* suggests although you could get away with having lease code in two places, we should think about refactoring it into its own class before we get to three copies.

Reply:
Ok, thank you for the comments. I'll resolve all of these and think about this last one.

Reply:
I saw that HttpNfcLease object returned from "ExportVApp", "ExportVm", "ImportVApp", vmware_deploy_ovf already use "ImportVApp" function, in this module using "ExportVm", one is upload, the other is download. I'm putting the common part of code in the module this time. If there is new requirement of "ExportVApp", then we can think about it.